### PR TITLE
Add RandomErasing for #263

### PR DESCRIPTION
**Pull request by mpskex**
Add RandomErasing for #263

### CODE REVIEW

**Comment by BloodAxe** [comment]
Hi! Could you please target branch of this PR to master?

**Comment by mpskex** [comment]
Sorry, just changed the base to master

File: albumentations/augmentations/transforms.py [hunk 1]
```diff
@@ -605,7 +606,104 @@ def get_transform_init_args_names(self):
         return ("mean", "std", "max_pixel_value")
 
 
-class Cutout(ImageOnlyTransform):
+class RandomErasing(DualTransform):
+    """ Randomly selects a rectangle region in an image and erases its pixels.
+        'Random Erasing Data Augmentation' by Zhong et al.
+        See https://arxiv.org/pdf/1708.04896.pdf
+    Args:
+         probability: The probability that the Random Erasing operation will be performed.
```
Comment (Dipet) [inline albumentations/augmentations/transforms.py:614]: `__init__` doesn't contain this parameter. It is named `p`

File: albumentations/augmentations/transforms.py [hunk 2]
```diff
@@ -605,7 +606,104 @@ def get_transform_init_args_names(self):
...
+         sl: Minimum proportion of erased area against input image.
+         sh: Maximum proportion of erased area against input image.
```
Comment (Dipet) [inline albumentations/augmentations/transforms.py:616]: What the difference between `sl` and `sh`?

File: albumentations/augmentations/transforms.py [hunk 3]
```diff
@@ -605,7 +606,104 @@ def get_transform_init_args_names(self):
...
+         r1: Minimum aspect ratio of erased area.
+         num_holes: Number of erasing areas
+         mean: Erasing value.
```
Comment (Dipet) [inline albumentations/augmentations/transforms.py:619]: `__init__` doesn't contain `mean` parameter

File: albumentations/augmentations/transforms.py [hunk 4]
```diff
@@ -605,7 +606,104 @@ def get_transform_init_args_names(self):
...
+        raise NotImplementedError(
+            'This part could be tricky. If you accidentally erased an object then it should be removed.')
+
+    def apply_to_keypoint(self, keypoint, holes=(), **params):
+        x, y, angle, scale = keypoint[:4]
+        for hx, hy, hw, hh in holes:
+            if x >= hx and x < hw + hx and y >= hy and y < hh + hy:
+                return None
+            else:
+                return keypoint
```
Comment (Dipet) [inline albumentations/augmentations/transforms.py:662]: With this condition for loop will work only once

File: albumentations/augmentations/transforms.py [hunk 5]
```diff
@@ -605,7 +606,104 @@ def get_transform_init_args_names(self):
...
+                # Fallback to central crop
+                w = int(round(math.sqrt(target_area / aspect_ratio)))
+                h = int(round(math.sqrt(target_area * aspect_ratio)))
+                x = int(round((img.shape[0] - w) / 2))
+                y = int(round((img.shape[0] - h) / 2))
+                holes.append((x, y, w, h))
+        return {'holes': holes}
+
+
+class Cutout(DualTransform):
```
Comment (Dipet) [inline albumentations/augmentations/transforms.py:706]: Why did you change `Cutout` base class to `DualTransform`?

Comment (mpskex) [reply albumentations/augmentations/transforms.py:706]: Sorry, that's a mistake.

File: albumentations/augmentations/transforms.py (hunk 4 above)
Comment (mpskex) [reply albumentations/augmentations/transforms.py:662]: will fix it.

File: albumentations/augmentations/transforms.py (hunk 3 above)
Comment (mpskex) [reply albumentations/augmentations/transforms.py:619]: The doc string is somehow not aligned. will be fixed.

File: albumentations/augmentations/transforms.py (hunk 2 above)
Comment (mpskex) [reply albumentations/augmentations/transforms.py:616]: The erased area will be in [sl * area, sh * area] per hole.

File: albumentations/augmentations/transforms.py (hunk 1 above)
Comment (mpskex) [reply albumentations/augmentations/transforms.py:614]: Parameter `probability` does not make its appearance in the actual `__init__` function. Will fix the docstring.

File: albumentations/augmentations/transforms.py (hunk 2 above)
Comment (Dipet) [reply albumentations/augmentations/transforms.py:616]: Just add this info into docstring.

File: albumentations/augmentations/transforms.py [hunk 6]
```diff
@@ -650,6 +651,103 @@ def get_transform_init_args_names(self):
...
+                 always_apply=False, p=0.5):
+        super(RandomErasing, self).__init__(always_apply=always_apply, p=p)
+        self.fill_value = fill_value
+        self.sl = sl
+        self.sh = sh
+        self.r1 = r1
+        self.num_holes = num_holes
+
+    def get_params(self):
+        return {}
```
Comment (Dipet) [inline albumentations/augmentations/transforms.py:678]: Please, remove unused function.

File: albumentations/augmentations/transforms.py [hunk 7]
```diff
@@ -650,6 +651,103 @@ def get_transform_init_args_names(self):
...
+
+    def get_transform_init_args_names(self):
+        return "sl", "sh", "r1", "num_holes", "fill_value"
+
+    def apply(self, img, holes=(), **kwargs):
+        for x, y, w, h in holes:
+            if img.size()[0] == 3:
+                img[0, x:x + w, y:y + h] = self.fill_value[0]
+                img[1, x:x + w, y:y + h] = self.fill_value[1]
+                img[2, x:x + w, y:y + h] = self.fill_value[2]
```
Comment (Dipet) [inline albumentations/augmentations/transforms.py:692]: We work with images in `[h, w, c]` wormat.
So change this lines to:
```python
img[x:x + w, y:y + h] = self.fill_value
```

File: albumentations/augmentations/transforms.py [hunk 8]
```diff
@@ -650,6 +651,103 @@ def get_transform_init_args_names(self):
...
+            else:
+                img[0, x:x + w, y:y + h] = self.fill_value[0]
+        return img
+
+    def apply_to_bbox(self, bbox, holes=(), **params):
+        # TODO: This part could be tricky. If you accidentally erased an object then it should be removed.
+        raise NotImplementedError(
```
Comment (Dipet) [inline albumentations/augmentations/transforms.py:699]: The default error message is fine. You don't need to implement `apply_to_bbox`

File: albumentations/augmentations/transforms.py [hunk 9]
```diff
@@ -650,6 +651,103 @@ def get_transform_init_args_names(self):
...
+            'This part could be tricky. If you accidentally erased an object then it should be removed.')
+
+    def apply_to_keypoint(self, keypoint, holes=(), **params):
+        x, y, angle, scale = keypoint[:4]
+        for hx, hy, hw, hh in holes:
+            if x >= hx and x < hw + hx and y >= hy and y < hh + hy:
+                return None
```
Comment (Dipet) [inline albumentations/augmentations/transforms.py:706]: Just simple change keypoint. Do not set it to `None`

File: albumentations/augmentations/transforms.py [hunk 10]
```diff
@@ -650,6 +651,103 @@ def get_transform_init_args_names(self):
...
+        return keypoint
+
+    def apply_to_keypoints(self, keypoints, **params):
+        kps = [self.apply_to_keypoint(tuple(keypoint[:4]), **params) + tuple(keypoint[4:]) for keypoint in keypoints]
```
Comment (Dipet) [inline albumentations/augmentations/transforms.py:710]: You don't need implement this method.

File: albumentations/augmentations/transforms.py [hunk 11]
```diff
@@ -650,6 +651,103 @@ def get_transform_init_args_names(self):
...
+        while None in kps:
+            kps.remove(None)
+        return kps
+
+    def apply_to_mask(self, img, holes=(), **params):
+        for x, y, w, h in holes:
```
Comment (Dipet) [inline albumentations/augmentations/transforms.py:716]: Just add the argument `mask_fill_value`. 
Look how it works in [CoarseDropout](https://github.com/albumentations-team/albumentations/blob/master/albumentations/augmentations/transforms.py#L721)

File: albumentations/augmentations/transforms.py [hunk 12]
```diff
@@ -650,6 +651,103 @@ def get_transform_init_args_names(self):
...
+            img[:, x:x + w, y:y + h] = 0
+        return img
+
+    def get_params_dependent_on_targets(self, params):
+        img = params["image"]
+        area = img.size()[1] * img.size()[2]
```
Comment (Dipet) [inline albumentations/augmentations/transforms.py:722]: We work with `np.ndarray` images. Use `img.shape`

**Review by Dipet — COMMENTED** [review]
Please, also add transform into these tests where `CoarseDropout` exists:
- https://github.com/albumentations-team/albumentations/blob/master/tests/test_serialization.py
- https://github.com/albumentations-team/albumentations/blob/master/tests/test_augmentations.py
- https://github.com/albumentations-team/albumentations/blob/master/tests/test_transforms.py

**Comment by Dipet** [comment]
A don't see a much difference from the current implementation of `CoarseDropout`
